core/commands/dag/dag.go

```diff
@@ -241,3 +261,259 @@ var DagResolveCmd = &cmds.Command{
+		// Accumulate any root CID seen in a header, and supplement its actual node if/when encountered
+		// We will attempt a pin *only* at the end in case all car files were well formed
+		//
+		// The boolean value indicates whether we have encountered the root within the car file's
+		expectedRootsSeen := make(map[cid.Cid]bool)
+
+		// grab a gc lock so that regardless of the size of the streamed in cars
+		// nothing will disappear on us before we had a chance to pin the roots
+		// at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+
+		// FIXME - none of these locker options work!
+		//
+		// This version deadlocks within this function, once you get to
+		// api.Pin().Add
+		//unlocker := node.Blockstore.GCLock()
+		//
+		// This version works in isolation but deadlocks with a concurrent GC run
+		// if I insert a time.Sleep(15 * time.Second) before the pin
+		unlocker := node.Blockstore.PinLock()
+		defer func() { unlocker.Unlock() }()
+
+		// this is *not* a transaction
+		// it is simply a way to relieve pressure on the blockstore
+		batch := ipld.NewBatch(req.Context, api.Dag())
+
+		it := req.Files.Entries()
+		for it.Next() {
+
+			file := files.FileFromEntry(it)
+			if file == nil {
+				return errors.New("expected a file handle")
+			}
+			defer func() { file.Close() }()
+
+			car, err := gocar.NewCarReader(file)
+			if err != nil {
+				return err
+			}
+
+			// Be explicit here, until the spec is finished
+			if car.Header.Version != 1 {
+				return errors.New("only car files version 1 supported at present")
+			}
+
+			for _, c := range car.Header.Roots {
+				if _, exists := expectedRootsSeen[c]; !exists {
+					expectedRootsSeen[c] = false
+				}
+			}
+
+			for {
+				block, err := car.Next()
+				if err != nil && err != io.EOF {
+					return err
+				} else if block == nil {
+					break
+				}
+
+				nd, err := ipld.Decode(block)
+				if err != nil {
+					return err
+				}
+
+				if err := batch.Add(req.Context, nd); err != nil {
+					return err
+				}
+
+				// encountered something known to be a root, for the first time
+				if seen, exists := expectedRootsSeen[nd.Cid()]; exists && !seen {
+					expectedRootsSeen[nd.Cid()] = true
+
+					if !silent {
+						if err := res.Emit(&OutputObject{Cid: nd.Cid()}); err != nil {
+							return err
+						}
+					}
+				}
+			}
+		}
+
+		if err := it.Err(); err != nil {
+			return err
+		}
+
+		if err := batch.Commit(); err != nil {
+			return err
+		}
+
+		if doPinRoots {
+			for c, seen := range expectedRootsSeen {
+				if !seen {
+					os.Stderr.WriteString("replace with seenfail emit\n")
+				}
+				// not ideal, but the pinning api takes only paths :(
+				rp := path.NewResolvedPath(
+					ipfspath.FromCid(c),
+					c,
+					c,
+					"",
+				)
+
+				if err := api.Pin().Add(req.Context, rp, options.Pin.Recursive(true)); err != nil {
```

Comment:
this is probably the deadlock. `api.Pin` takes the pin lock internally so we don't GC while validating that we actually have the DAG.

We have to call the _non_ core API pinner (ipfsnode.Pinner), unfortunately.

Note: to debug things like this, it can be useful to take a stack dump with pprof.